tango_simlib/tango_sim_generator.py

```diff
@@ -282,63 +299,80 @@ def initialize_dynamic_attributes(self):
-                attr_dtype = meta_data["data_type"]
-                d_format = meta_data["data_format"]
                 # Dynamically add all attributes except those with DevEnum data type,
                 # and SPECTRUM data format since they are added statically to the device
                 # class prior to start-up. Also exclude attributes with a data format
                 # 'IMAGE' as we currently do not handle them.
-                if str(attr_dtype) == "DevEnum":
+                if not self._is_attribute_addable_dynamically(meta_data):
                     continue
-                elif str(d_format) == "SPECTRUM":
+                # The return value of rwType is a string and it is required as a
+                # PyTango data type when passed to the Attr function.
+                # e.g. 'READ' -> tango._tango.AttrWriteType.READ
+                rw_type = meta_data["writable"]
+                rw_type = getattr(AttrWriteType, rw_type)
+                attr = self._create_attribute(
+                    attribute_name, meta_data["data_type"], rw_type
+                )
+                if attr is None:
                     continue
-                elif str(d_format) == "IMAGE":
-                    self._not_added_attributes.append(attribute_name)
+
+                self._configure_attribute_default_properties(attr, meta_data)
+                self._add_dynamic_attribute(attr, rw_type)
+                MODULE_LOGGER.info("Added dynamic {} attribute".format(attribute_name))
+
+        def _add_dynamic_attribute(self, attribute, read_write_type):
+            if read_write_type in (AttrWriteType.READ, AttrWriteType.READ_WITH_WRITE):
+                self.add_attribute(attribute, r_meth=self.read_attributes)
+            elif read_write_type == AttrWriteType.WRITE:
+                self.add_attribute(attribute, w_meth=self.write_attributes)
+            elif read_write_type == AttrWriteType.READ_WRITE:
+                self.add_attribute(
+                    attribute, r_meth=self.read_attributes, w_meth=self.write_attributes
+                )
+
+        def _is_attribute_addable_dynamically(self, quantity_meta_data):
+            attr_dtype = quantity_meta_data["data_type"]
+            d_format = quantity_meta_data["data_format"]
+            if str(attr_dtype) == "DevEnum" or str(d_format) == "SPECTRUM":
+                return False
+            elif str(d_format) == "IMAGE":
+                self._not_added_attributes.append(quantity_meta_data["name"])
+                return False
+
+            return True
+
+        def _create_attribute(self, attribute_name, attr_dtype, rw_type):
+            attribute = None
+            # Add a try/except clause when creating an instance of Attr class
+            # as PyTango may raise an error when things go wrong.
+            try:
+                attribute = Attr(attribute_name, attr_dtype, rw_type)
+                return attribute
```

Comment:
Returning `attribute` can be added to a `finally` block rather than doing it twice